spmd/tensor/api.py

```diff
@@ -12,6 +12,62 @@
 """
 _DEBUG_STRICT = False
 
+# NOTE [Autograd interaction between torch.Tensor]
+#
+# The autograd functions defined below are being used by the public
+# facing APIs (i.e. from_local, local_tensor) to ensure our Tensor
+# works together with torch.Tensor within autograd engine. This
+# allows DistributedTensor to exist on part of the module hierarchy
+# and still able to calculate gradients across the torch.Tensor and
+# DistributedTensor boundary.
+# As an example, it enables backward computation of the following:
+#   Module A (with all torch.Tensor params)
+#       -> Module B (with all DistributedTensor params)
+#           -> Module C (with all torch.Tensor params)
+class ToTorchTensor(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input: "Tensor"):  # type: ignore
+        ctx.previous_placement = input.placements
+        ctx.previous_device_mesh = input.device_mesh
+        return input._local_tensor
+
+    @staticmethod
+    def backward(ctx, grad_output: torch.Tensor):  # type: ignore
+        previous_placement = ctx.previous_placement
+        previous_device_mesh = ctx.previous_device_mesh
+        dist_grad = Tensor.from_local(
+            grad_output, previous_device_mesh, previous_placement
+        )
+        return dist_grad
+
+
+class FromTorchTensor(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input: torch.Tensor, device_mesh, placements):  # type: ignore
+        ctx.previous_placement = placements
+        ctx.previous_device_mesh = device_mesh
+
+        dist_tensor = Tensor(
+            input,
+            device_mesh,
+            placements,
+        )
+        return dist_tensor
+
+    @staticmethod
+    def backward(ctx, grad_output: "Tensor"):  # type: ignore
+        previous_placement = ctx.previous_placement
+        previous_device_mesh = ctx.previous_device_mesh
+
+        # reshard to the placement when creating DistributedTensor
+        # so that the gradient layout matches, and we could return
+        # local gradients directly
+        if grad_output.placements != previous_placement:
+            grad_output = grad_output.redistribute(
+                previous_device_mesh, previous_placement
+            )
+        return grad_output._local_tensor, None, None
+
 
 class Tensor(torch.Tensor):  # pyre-ignore[13]: pyre is bad at __new__
```

Comment:
Why not `DistributedTensor`? That makes the code pretty hard to read in this file :/
Also it makes the typing very hard to follow as everything is typed "Tensor" !

Reply:
At some point, we probably move the whole DistributedTensor implementation to core torch.distributed, and there was a discussion on whether we should call it a `DistributedTensor` or other names. One argument is that we don't want the naming ended be `torch.distributed.DistributedTensor`, as there're duplicate "distributed" there, `torch.distributed.Tensor` looks better, and that's the reason we are having it as "Tensor" now.  Although I agree it's a bit confusing..

Any suggestions on naming? Also cc @aazzolini @yhcharles @wangkuiyi @jamesr66a @mrshenli on bettering naming of DistributedTensor. 

Or we can do as follows: This tensor subclass definition we switch to use `DistributedTensor`. But when we expose it in package init (i.e. in torch.distributed.__init__.py), we can do things like following:

`from spmd.tensor.api import DistributedTensor as Tensor`

This will make an name alias for DistributedTensor, and still appear from the package level as `torch.distributed.Tensor`

Reply:
Strongly agree re. code readability issues. I am supportive of renaming to DistributedTensor and if it's clear from the context user can always alias it.

Reply:
`DTensor` maybe?

Reply:
> `DTensor` maybe?

Yeah that's actually a good one! Short enough to distinguish with torch.Tensor, and even if we have `torch.distributed.DTensor` it looks still reasonable! This also seems consistent with [TF's DTensor](https://www.tensorflow.org/guide/dtensor_overview) and easy to map between each other. 

I am down with `DTensor` :)

Reply:
It's happening here https://github.com/pytorch/PiPPy/pull/330